src/storage/mvcc/reader/reader.rs

```diff
@@ -557,29 +559,65 @@ impl<S: EngineSnapshot> MvccReader<S> {
         if !ok {
             return Ok((vec![], false));
         }
-        let mut writes = Vec::with_capacity(limit);
+        let mut cur_key = None;
+        let mut key_old_writes = Vec::with_capacity(limit);
         while cursor.valid()? {
             let key = Key::from_encoded_slice(cursor.key(&mut self.statistics.write));
             if let Some(end) = end {
                 if key >= *end {
-                    return Ok((writes, false));
+                    return Ok((key_old_writes, false));
                 }
             }
-
-            if filter(&key) {
-                writes.push((
-                    key,
-                    WriteRef::parse(cursor.value(&mut self.statistics.write))?.to_owned(),
-                ));
-                if limit > 0 && writes.len() == limit {
-                    return Ok((writes, true));
+            let commit_ts = key.decode_ts()?;
+            let truncated_key = key.truncate_ts()?;
+            // 1. To make sure we only check each unique key once.
+            // 2. No need to find an old version for the key if its latest `commit_ts` is
+            // smaller than or equal to the version.
+            if cur_key.is_some() && cur_key.clone().unwrap() == truncated_key
+                || cur_key.is_none() && commit_ts <= version
+            {
+                cursor.next(&mut self.statistics.lock);
+                continue;
+            }
+            cur_key = Some(truncated_key.clone());
+
+            let mut old_write = None;
+            let old_version_key = truncated_key.clone().append_ts(version);
+            // Try to seek to the key with the specified version.
+            if cursor.near_seek(&old_version_key, &mut self.statistics.write)?
+                && Key::is_user_key_eq(
+                    cursor.key(&mut self.statistics.write),
+                    truncated_key.as_encoded(),
+                )
+            {
+                while cursor.valid()? {
+                    old_write =
+                        Some(WriteRef::parse(cursor.value(&mut self.statistics.write))?.to_owned());
+                    // Move to the next key.
+                    cursor.next(&mut self.statistics.lock);
+                    match old_write.as_ref().unwrap().write_type {
+                        WriteType::Put | WriteType::Delete => {
+                            break;
+                        }
+                        WriteType::Lock | WriteType::Rollback => {
+                            // We should find the latest visible version after it.
+                            let key =
+                                Key::from_encoded_slice(cursor.key(&mut self.statistics.write));
+                            if key.truncate_ts()? != truncated_key {
+                                break;
```

Comment:
Need to set `old_write` and `cur_key` to `None` when the key is mismatch? 
`cur_key` is never reset to `None` in this function, I'm think the test coverage is not enough,

Reply:
Nice catch.